### PR TITLE
Extend meta_df tests to include datetime

### LINKED ISSUE #222: Aggregate fails with time column
**Describe the bug**

If you make a dataframe with datetime times and call its `aggregate` method, you receive an error. This should work the same as a dataframe with year columns. 

```python
>>> from datetime import datetime
>>> 
>>> import pandas as pd
>>> 
>>> from pyam import IamDataFrame
>>> 
>>> eg_df = pd.DataFrame([
...     ['model_a', 'scen_a', 'World', 'Primary Energy|Gas', 'EJ/y', 1, 6.],
...     ['model_a', 'scen_a', 'World', 'Primary Energy|Coal', 'EJ/y', 0.5, 3],
...     ['model_a', 'scen_b', 'World', 'Primary Energy|Gas', 'EJ/y', 2, 7],
... ],
...     columns=['model', 'scenario', 'region', 'variable', 'unit', 2005, 2010],
... )
>>> year_df = IamDataFrame(data=eg_df)
>>> year_df.timeseries()
                                                  2005  2010
model   scenario region variable            unit            
model_a scen_a   World  Primary Energy|Coal EJ/y   0.5   3.0
                        Primary Energy|Gas  EJ/y   1.0   6.0
        scen_b   World  Primary Energy|Gas  EJ/y   2.0   7.0
>>> 
>>> year_df.aggregate('Primary Energy', append=True)
>>> year_df.timeseries()
                                                  2005  2010
model   scenario region variable            unit            
model_a scen_a   World  Primary Energy      EJ/y   1.5   9.0
                        Primary Energy|Coal EJ/y   0.5   3.0
                        Primary Energy|Gas  EJ/y   1.0   6.0
        scen_b   World  Primary Energy      EJ/y   2.0   7.0
                        Primary Energy|Gas  EJ/y   2.0   7.0
>>> 
>>> dt_df = IamDataFrame(data=eg_df.rename(
...     {2005: datetime(2005, 1, 1), 2010: datetime(2010, 6, 1)},
...     axis="columns"
... ))
>>> dt_df.timeseries()
                                                  2005-01-01  2010-06-01
model   scenario region variable            unit                        
model_a scen_a   World  Primary Energy|Coal EJ/y         0.5         3.0
                        Primary Energy|Gas  EJ/y         1.0         6.0
        scen_b   World  Primary Energy|Gas  EJ/y         2.0         7.0
>>> 
>>> dt_df.aggregate('Primary Energy', append=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/zebedeenicholls/Documents/AGCEC/Misc/pyam-zn-fork/pyam/core.py", line 700, in aggregate
    df_components = _aggregate_by_variables(self.data, components, units)
  File "/Users/zebedeenicholls/Documents/AGCEC/Misc/pyam-zn-fork/pyam/core.py", line 1319, in _aggregate_by_variables
    return df.groupby(GROUP_IDX).sum()['value']
  File "/Users/zebedeenicholls/miniconda3/envs/pyam-zn-fork/lib/python3.6/site-packages/pandas/core/generic.py", line 7632, in groupby
    observed=observed, **kwargs)
  File "/Users/zebedeenicholls/miniconda3/envs/pyam-zn-fork/lib/python3.6/site-packages/pandas/core/groupby/groupby.py", line 2110, in groupby
    return klass(obj, by, **kwds)
  File "/Users/zebedeenicholls/miniconda3/envs/pyam-zn-fork/lib/python3.6/site-packages/pandas/core/groupby/groupby.py", line 360, in __init__
    mutated=self.mutated)
  File "/Users/zebedeenicholls/miniconda3/envs/pyam-zn-fork/lib/python3.6/site-packages/pandas/core/groupby/grouper.py", line 578, in _get_grouper
    raise KeyError(gpr)
KeyError: 'year'
```

**Failing Test**

Change `meta_df` to `test_df` in all tests in `tests/test_feature_aggregate.py`.

**System (please complete the following information):**

 - OS: macOS
 - Python and pyam version: Python 3.6.7, 0.1.2+53.gb0e6562.dirty

**Additional context**

The failure occurs because the definition of `GROUP_IDX` is static. Hence line 1319 of `pyam/core.py` looks for a `year` column and doesn't find it, hence the failure. I think there are two possible solutions here:

1. Like was done with `LONG_IDX`, move `GROUP_IDX` to being an attribute of `IamDataFrame`. It would make sense to also move `YEAR_IDX`, `REGION_IDX` and `SORT_IDX` at the same time in order to avoid these failures in other places. (I would suggest moving all indexes to being attributes of `IamDataFrame` for consistency.)
1. Drop `year` support and always use `time` columns. Given earlier discussions I think this one is unlikely to happen, but this instance is another place where it would be a sensible solution. Just for completeness, I think dropping `year` support has a couple of extra arguments in its favour. It's always possible to recover year information from a datetime. It would make the meaning of the `time` column uniform (always being a datetime) and hence force users with other `time` information to put it in more well named columns (e.g. `Summer evening` might have to go in a `season-time-of-day` column instead to be more explicit, it would be annoying to try and filter e.g. `time` based on year only to have it explode because `time` is full of strings).

### PR DESCRIPTION
**Pull request by znicholls**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added (N/A)
- [x] Description in RELEASE_NOTES.md Added

## Adding to RELEASE_NOTES.md (remove section after adding to RELEASE_NOTES.md)

Please add a single line in the release notes similar to the following:

```
- (#XX)[http://link-to-pr.com] Added feature which does something
```

# Description of PR

Closes #222. Good news is this was already sorted in v0.2.0 so all this does is make sure it stays that way by expanding the tests a bit.

### CODE REVIEW

File: pyam/core.py [hunk 1]
```diff
@@ -309,6 +309,31 @@ def interpolate(self, year):
...
+        ------
+        ValueError
+            "time" is not a column of `self.data`
+        """
+        if "time" not in self.data:
+            raise ValueError("time column must be datetime to use this method")
+
+        ret = copy.deepcopy(self) if not inplace else self
+
+        ret.data["year"] = ret.data["time"].apply(lambda x: x.year)
```
Comment (danielhuppmann) [inline pyam/core.py]: what would happen here if you have two `time` entries for the same year - if I understand this correctly, the resulting `IamDataFrame` would then have a non-unique `year` in `data` for the same model/scenario/variable/region?

This does not cause an issue in the new unit tests, but maybe it makes sense to add a sanity check here similar to https://github.com/IAMconsortium/pyam/blob/25aa7f09fb49ed64569255dd2bf260fb8e62f717/pyam/utils.py#L252

File: tests/test_feature_append_rename_convert.py [hunk 2]
```diff
@@ -160,15 +166,20 @@ def test_rename_append(meta_df):
     obs = meta_df.rename(mapping, append=True)
 
     # test data changes
+    dts = [dt.datetime(2005, 6, 17), dt.datetime(2010, 7, 21)]
+    times = [2005, 2010] if "year" in meta_df.data else dts
```
Comment (danielhuppmann) [inline tests/test_feature_append_rename_convert.py]: why double-quotes?

File: tests/conftest.py [hunk 3]
```diff
@@ -195,9 +195,17 @@ def test_pd_df():
...
 
 
-@pytest.fixture(scope="function")
-def meta_df():
-    df = IamDataFrame(data=TEST_DF)
+@pytest.fixture(
+    scope="function",
+    params=[
+        [2005, 2010],
+        [datetime(2005, 6, 17), datetime(2010, 7, 21)],
```
Comment (danielhuppmann) [inline tests/conftest.py]: this list of `datetimes` is used quite a lot now throughout the tests - maybe make that available as static object and reuse throughout?

File: tests/test_cast_to_iamc.py [hunk 4]
```diff
@@ -58,17 +67,21 @@ def test_cast_with_model_arg(meta_df):
 
 
 def test_cast_by_column_concat(meta_df):
+    dts = [dt.datetime(2005, 6, 17), dt.datetime(2010, 7, 21)]
```
Comment (danielhuppmann) [inline tests/test_cast_to_iamc.py]: reuse from `test/conftest.py`

File: tests/test_core.py [hunk 5]
```diff
@@ -518,22 +518,39 @@ def test_validate_up(meta_df):
...
 
 def test_validate_both(meta_df):
     obs = meta_df.validate({'Primary Energy': {'up': 6.5, 'lo': 2.0}})
     assert len(obs) == 2
-    assert list(obs['year'].values) == [2005, 2010]
+    if 'year' in meta_df.data:
+        assert list(obs['year'].values) == [2005, 2010]
+    else:
+        exp_time = pd.to_datetime([
+            datetime.datetime(2005, 6, 17),
```
Comment (danielhuppmann) [inline tests/test_core.py]: reuse from `test/conftest.py`

File: tests/test_feature_aggregate.py [hunk 6]
```diff
@@ -9,13 +12,19 @@ def test_do_aggregate_append(meta_df):
     meta_df.aggregate('Primary Energy', append=True)
     obs = meta_df.filter(variable='Primary Energy').timeseries()
 
+    dts = [dt.datetime(2005, 6, 17), dt.datetime(2010, 7, 21)]
```
Comment (danielhuppmann) [inline tests/test_feature_aggregate.py]: reuse from `test/conftest.py`

File: tests/test_feature_append_rename_convert.py [hunk 7]
```diff
@@ -136,13 +137,18 @@ def test_rename_index(meta_df):
     obs = meta_df.rename(mapping, scenario={'scen_a': 'scen_c'})
 
     # test data changes
+    dts = [dt.datetime(2005, 6, 17), dt.datetime(2010, 7, 21)]
```
Comment (danielhuppmann) [inline tests/test_feature_append_rename_convert.py]: reuse from `test/conftest.py`

File: tests/test_feature_compare.py [hunk 8]
```diff
@@ -13,17 +16,23 @@ def test_compare(meta_df):
...
-        columns=['variable', 'unit', 'year', 'meta_df', 'clone'],
+        columns=['variable', 'unit', 'time', 'meta_df', 'clone'],
     )
     exp['model'] = 'model_a'
     exp['scenario'] = 'scen_a'
     exp['region'] = 'World'
-    exp = exp.set_index(IAMC_IDX + ['year'])
+    time_col = 'time'
+    if "year" in meta_df.data.columns:
+        exp["year"] = exp["time"].apply(lambda x: x.year)
```
Comment (danielhuppmann) [inline tests/test_feature_compare.py]: use the new function `swap_time_for_year()`?

File: pyam/core.py (hunk 1 above)
Comment (znicholls) [reply pyam/core.py]: Nice will do

File: tests/test_feature_append_rename_convert.py (hunk 2 above)
Comment (znicholls) [reply tests/test_feature_append_rename_convert.py]: habit, have changed

File: tests/test_feature_compare.py (hunk 8 above)
Comment (znicholls) [reply tests/test_feature_compare.py]: can't do that here as `exp` has nan's so can't be cast to `IamDataFrame`

File: tests/conftest.py (hunk 3 above)
Comment (znicholls) [reply tests/conftest.py]: nice will do!

File: tests/test_feature_compare.py (hunk 8 above)
Comment (danielhuppmann) [reply tests/test_feature_compare.py]: got it, thanks!

**Comment by znicholls** [comment]
cheers @danielhuppmann, that should all be addressed now!

**Comment by znicholls** [comment]
not sure what's going on with CI, read through https://github.com/pyproj4/pyproj/issues/134 but am none the wiser...

**Comment by danielhuppmann** [comment]
sorry about this - @gidden and I already discussed that the region-plotting-tests are quite unstable. How about deactivating those for Windows machines?

**Comment by gidden** [comment]
Hi folks - indeed these are our most finicky tests. One suggestion before turning them off for windows.

In pyproj4/pyproj#134, someone notes that the issue goes away for pyproj>2.1.

On appveyor, I see that the version is:
```
pyproj: 1.9.6
```

So perhaps we can try adding the minimum version to our appveyor install? 

If that still fails, then I'm ok with just turning them off on windows..

**Comment by gidden** [comment]
Hey @znicholls, I'm now digging through more geopandas/cartopy/pyproj issues. I have two thoughts.

First is related to geopandas/geopandas#830, which suggests a specific pyproj version for windows. To attempt this, can you please revert the update in ` ci/environment-conda-forge.txt` and try pinning *only* the appveyor install to `"pyproj=1.9.4"`?

Second, I'm looking through our last successful CI on windows: https://ci.appveyor.com/project/gidden/pyam/builds/24523793/job/2idsdm0hhgvg2tfw

In reviewing the package versions at runtime, there is no difference between that run and 60ccc9a. In reviewing the installed conda packages, however, I see the following difference:

Successful run:

```
proj4                     5.2.0                ha925a31_1 
```

Failed run:
```
proj4                     5.2.0             h6538335_1002    conda-forge
```

Thus this implies one potential fix is to install `proj4` via the first `conda install` (without `-c conda-forge`).

I know this is annoying, but if you would be willing to try those two changes, that would be amazing. Otherwise, let's just turn this stuff off for windows.

**Comment by gidden** [comment]
I don't know whether to be happy or sad, but it looks fixed now!

File: pyam/core.py [hunk 9]
```diff
@@ -309,6 +309,36 @@ def interpolate(self, year):
...
+
+        Raises
+        ------
+        ValueError
+            "time" is not a column of `self.data`
+        """
+        if "time" not in self.data:
+            raise ValueError("time column must be datetime to use this method")
+
+        ret = copy.deepcopy(self) if not inplace else self
```
Comment (gidden) [inline pyam/core.py]: should this now simply be `self.copy()` since we have that method?

File: pyam/core.py [hunk 10]
```diff
@@ -1358,6 +1388,10 @@ def _check_rows(rows, check, in_range=True, return_test='any'):
         msg = 'Unknown checking type: {}'
         raise ValueError(msg.format(check.keys() - valid_checks))
 
+    rows = rows.copy()
```
Comment (gidden) [inline pyam/core.py]: should this copy move into the `if`statement? otherwise we might be doing a copy unnecessarily?

**Comment by gidden** [comment]
A few inline comments, then this is g2g for me. @danielhuppmann you still have this blocking in review, so let us know if it passes muster!

File: pyam/core.py (hunk 10 above)
Comment (znicholls) [reply pyam/core.py]: It's to ensure you don't accidentally add  a `year` column to your dataframe

Comment (gidden) [reply pyam/core.py]: But wouldn't that only happen if you get inside the `if` statement? Suggestion would be:

```
    if 'year' not in rows:
        rows = rows.copy()
        rows['year'] = rows['time'].apply(lambda x: x.year)
```

Comment (gidden) [reply pyam/core.py]: Or perhaps I am confused. Sorry if that's the case!

Comment (znicholls) [reply pyam/core.py]: ah yep I get it, nice!

**Comment by gidden** [comment]
Looks great! I'll let @danielhuppmann pull this one in.

**Pull request merged**